test/Kestrel.FunctionalTests/Http2/ShutdownTests.cs

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#if NETCOREAPP2_2
+
+using System.Collections.Generic;
+using System.Net;
+using System.Net.Http;
+using System.Runtime.InteropServices;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Server.Kestrel.Core;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2;
+using Microsoft.AspNetCore.Testing;
+using Microsoft.AspNetCore.Testing.xunit;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.FunctionalTests.Http2
+{
+    [OSSkipCondition(OperatingSystems.MacOSX, SkipReason = "Missing SslStream ALPN support: https://github.com/dotnet/corefx/issues/30492")]
+    [OSSkipCondition(OperatingSystems.Linux, SkipReason = "Curl requires a custom install to support HTTP/2, see https://askubuntu.com/questions/884899/how-do-i-install-curl-with-http2-support")]
+    [MinimumOSVersion(OperatingSystems.Windows, WindowsVersions.Win10)]
+    public class ShutdownTests : TestApplicationErrorLoggerLoggedTest
+    {
+        private static X509Certificate2 _x509Certificate2 = TestResources.GetTestCertificate();
+
+        private HttpClient Client { get; set; }
```

Comment:
I think it would be better to model these tests after TlsTests which doesn't use HttpClient but instead is written against the raw connection using the Http2Frame type. I would even consider doing this test without SSL so the tests can run on all platforms. This will also make it easier to switch to an in-memory test.

Reply:
I don't mind having an integration test for a scenario like shutdown to make sure a real client can get a graceful response before the connection closes.

Reply:
For what possible reason would a real client be different than a fake one in this scenario? What's important is how the server behaves when the app is still running when shutdown starts. I don't mind having _some_ integration tests specifically for making sure HTTP/2 works with a real client, but when we're testing specific things we should write the tests in a way that will allow them to run with our in-memory test transport.

Reply:
This is one of those cases where we'd want to make sure our shutdown behavior matches the client's expectations. E.g. that sending multiple go-aways is handled as expected, and that the response gets fully flushed before the connection closes.

Reply:
Fair enough 👍 . In the future we might want to think about moving tests that don't test the transport itself, but do require a real client, in a separate functional test project that doesn't run per transport.